zmq_addon.hpp

```diff
@@ -40,27 +40,25 @@ namespace zmq
 
 #ifdef ZMQ_CPP11
 
-/*  Receive a multipart message.
-    
-    Writes the zmq::message_t objects to OutputIterator out.
-    The out iterator must handle an unspecified number of writes,
-    e.g. by using std::back_inserter.
-    
-    Returns: the number of messages received or nullopt (on EAGAIN).
-    Throws: if recv throws. Any exceptions thrown
-    by the out iterator will be propagated and the message
-    may have been only partially received with pending
-    message parts. It is adviced to close this socket in that event.
-*/
-template<class OutputIt>
-ZMQ_NODISCARD
-recv_result_t recv_multipart(socket_ref s, OutputIt out,
-                             recv_flags flags = recv_flags::none)
+namespace detail
+{
+template<bool CheckN, class OutputIt>
+recv_result_t recv_multipart_n(socket_ref s, OutputIt out, size_t n,
+                               recv_flags flags)
 {
     size_t msg_count = 0;
     message_t msg;
     while (true)
     {
+        #ifdef ZMQ_CPP17
+        if constexpr (CheckN)
+        #else
+        if (CheckN)
+        #endif
```

Comment:
One minor comment on this: I think it is not worthwhile to conditionally use `constexpr` here.

While I really appreciate `if constexpr`, I think of it primarily as a means to simplify code, e.g. where previously template specializations or multiple SFINAE overloads needed to be used. It is necessary when the code in the branch is not valid when the condition is false. This is not the case here. The compiler is easily able to optimize the if away if the condition is false without `constexpr`. Here, due to the necessity of the `#ifdef`, the code is more complex with the `constexpr` than without, so I would rather leave it out.

Reply:
Yes, good point.